File: opendc-integration-jpa/core/src/main/kotlin/nl/atlarge/opendc/integration/jpa/schema/Rack.kt
/*
 * MIT License
 *
 * Copyright (c) 2017 atlarge-research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package nl.atlarge.opendc.integration.jpa.schema

import nl.atlarge.opendc.topology.container.Rack
import javax.persistence.Entity

/**
 * A rack entity in a room in the persistent schema.
 *
 * @property id The unique identifier of the rack.
 * @property name The name of the rack.
 * @property capacity The capacity of the rack in terms of units.
 * @property powerCapacity The power capacity of the rack in Watt.
 * @property machines The machines in the rack.
 * @author Fabian Mastenbroek (f.s.mastenbroek@student.tudelft.nl)
 */
@Entity
class Rack(
	id: Int,
	val name: String,
	val capacity: Int,
	val powerCapacity: Int,
	val machines: List<Machine>
): RoomObject(id), Rack {
	/**
	 * The initial state of the entity.
	 */
	override val initialState = Unit
}

File: opendc-web/opendc-web-server/src/main/webui/redux/reducers/topology/rack.js
/*
 * Copyright (c) 2021 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import produce from 'immer'
import { STORE_TOPOLOGY } from '../../actions/topology'
import { DELETE_MACHINE } from '../../actions/topology/machine'
import { DELETE_RACK, EDIT_RACK_NAME, ADD_MACHINE } from '../../actions/topology/rack'
import { ADD_RACK_TO_TILE } from '../../actions/topology/room'

function rack(state = {}, action, { machines }) {
    switch (action.type) {
        case STORE_TOPOLOGY:
            return action.entities.racks || {}
        case ADD_RACK_TO_TILE:
            return produce(state, (draft) => {
                const { rack } = action
                draft[rack.id] = rack
            })
        case EDIT_RACK_NAME:
            return produce(state, (draft) => {
                const { rackId, name } = action
                draft[rackId].name = name
            })
        case DELETE_RACK:
            return produce(state, (draft) => {
                const { rackId } = action
                delete draft[rackId]
            })
        case ADD_MACHINE:
            return produce(state, (draft) => {
                const { machine } = action
                draft[machine.rackId].machines.push(machine.id)
            })
        case DELETE_MACHINE:
            return produce(state, (draft) => {
                const { machineId } = action
                const machine = machines[machineId]
                const rack = draft[machine.rackId]
                const index = rack.machines.indexOf(machineId)
                rack.machines.splice(index, 1)
            })
        default:
            return state
    }
}

export default rack
